File: 238-product-of-array-except-self/product-of-array-except-self.java
class Solution {
    public int[] productExceptSelf(int[] nums) {
      int n = nums.length;
      int[] left = new int[n];
      int[] right = new int[n];

      int l = 1 , r = 1;

      for(int i = 0 ; i < n ; i++){
        left[i] = l;
        l *= nums[i];
      }  

      for(int i = n-1 ; i >= 0 ; i--){
        right[i] = r;
        r *= nums[i];
      }

      int[] res = new int[n];
      for(int i = 0 ; i < n ; i++){
        res[i] = left[i]*right[i];
      }

      return res;
    }
}

File: 238-product-of-array-except-self/product-of-array-except-self.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:

        left = []
        l = 1
        n = len(nums)

        for i in range(n):
            left.append(l)
            l *= nums[i]
        
        right = []
        r = 1
        for i in range(n-1 , -1 , -1):
            left[i] *= r
            r *= nums[i]
        
        return left